aws/resource_aws_sagemaker_notebook_instance_test.go

```diff
@@ -0,0 +1,324 @@
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckSagemakerNotebookInstanceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccSagemakerNotebookInstanceConfig(notebookName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckSagemakerNotebookInstanceExists(resourceName, &notebook),
+
+					resource.TestCheckResourceAttr(
+						"aws_sagemaker_notebook_instance.foo", "instance_type", "ml.t2.medium"),
+				),
+			},
+
+			{
+				Config: testAccSagemakerNotebookInstanceUpdateConfig(notebookName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckSagemakerNotebookInstanceExists("aws_sagemaker_notebook_instance.foo", &notebook),
+
+					resource.TestCheckResourceAttr(
+						"aws_sagemaker_notebook_instance.foo", "instance_type", "ml.m4.xlarge"),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAccAWSSagemakerNotebookInstance_tags(t *testing.T) {
+	var notebook sagemaker.DescribeNotebookInstanceOutput
+	notebookName := resource.PrefixedUniqueId(sagemakerTestAccSagemakerNotebookInstanceResourceNamePrefix)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckSagemakerNotebookInstanceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccSagemakerNotebookInstanceTagsConfig(notebookName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckSagemakerNotebookInstanceExists("aws_sagemaker_notebook_instance.foo", &notebook),
+					testAccCheckSagemakerNotebookInstanceTags(&notebook, "foo", "bar"),
+
+					resource.TestCheckResourceAttr(
+						"aws_sagemaker_notebook_instance.foo", "name", notebookName),
+					resource.TestCheckResourceAttr("aws_sagemaker_notebook_instance.foo", "tags.%", "1"),
+					resource.TestCheckResourceAttr("aws_sagemaker_notebook_instance.foo", "tags.foo", "bar"),
+				),
+			},
+
+			{
+				Config: testAccSagemakerNotebookInstanceTagsUpdateConfig(notebookName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckSagemakerNotebookInstanceExists("aws_sagemaker_notebook_instance.foo", &notebook),
+					testAccCheckSagemakerNotebookInstanceTags(&notebook, "foo", ""),
+					testAccCheckSagemakerNotebookInstanceTags(&notebook, "bar", "baz"),
+
+					resource.TestCheckResourceAttr("aws_sagemaker_notebook_instance.foo", "tags.%", "1"),
+					resource.TestCheckResourceAttr("aws_sagemaker_notebook_instance.foo", "tags.bar", "baz"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckSagemakerNotebookInstanceDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).sagemakerconn
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_sagemaker_notebook_instance" {
+			continue
+		}
+
+		resp, err := conn.ListNotebookInstances(&sagemaker.ListNotebookInstancesInput{
```

Comment:
Should this be using `DescribeNotebookInstance` to match the `Read` function and `Exists` test function? A name contains check seems more flakey than directly checking.